File: steps/javascript/deletion-reaper.cron.step.js
// steps/javascript/deletion-reaper.cron.step.js
const { findDeletedPetsReadyToPurge, remove } = require('./js-store');

exports.config = {
  type: 'cron',
  name: 'JsDeletionReaper',
  description: 'Daily job that permanently removes pets scheduled for deletion',
  cron: '0 2 * * *', // Daily at 2:00 AM
  emits: [],
  flows: ['JsPetManagement']
};

exports.handler = async ({ emit, logger }) => {
  if (logger) {
    logger.info('🔄 Deletion Reaper started - scanning for pets to purge');
  }

  try {
    const petsToReap = findDeletedPetsReadyToPurge();
    
    if (petsToReap.length === 0) {
      if (logger) {
        logger.info('✅ Deletion Reaper completed - no pets to purge');
      }
      
      if (emit) {
        await emit({
          topic: 'js.reaper.completed',
          data: { 
            scannedAt: Date.now(),
            purgedCount: 0,
            message: 'No pets ready for purging'
          }
        });
      }
      return;
    }

    let purgedCount = 0;
    
    for (const pet of petsToReap) {
      const success = remove(pet.id);
      
      if (success) {
        purgedCount++;
        
        if (logger) {
          logger.info('💀 Pet permanently purged', { 
            petId: pet.id, 
            name: pet.name,
            deletedAt: new Date(pet.deletedAt).toISOString(),
            purgeAt: new Date(pet.purgeAt).toISOString()
          });
        }

        if (emit) {
          await emit({
            topic: 'js.pet.purged',
            data: { 
              petId: pet.id, 
              name: pet.name,
              species: pet.species,
              deletedAt: pet.deletedAt,
              purgedAt: Date.now()
            }
          });
        }
      } else {
        if (logger) {
          logger.warn('⚠️ Failed to purge pet', { petId: pet.id, name: pet.name });
        }
      }
    }

    if (logger) {
      logger.info('✅ Deletion Reaper completed', { 
        totalScanned: petsToReap.length,
        purgedCount,
        failedCount: petsToReap.length - purgedCount
      });
    }

    if (emit) {
      await emit({
        topic: 'js.reaper.completed',
        data: { 
          scannedAt: Date.now(),
          purgedCount,
          totalScanned: petsToReap.length
        }
      });
    }

  } catch (error) {
    if (logger) {
      logger.error('❌ Deletion Reaper error', { error: error.message });
    }
  }
};

File: src/python/deletion_reaper.cron_step.py
# src/python/deletion_reaper.cron.step.py
config = {
    "type": "cron",
    "name": "PyDeletionReaper",
    "description": "Daily job that permanently removes pets scheduled for deletion",
    "cron": "0 2 * * *",  # Daily at 2:00 AM
    "emits": [],
    "flows": ["PyPetManagement"]
}

async def handler(ctx):
    logger = getattr(ctx, 'logger', None) if ctx else None
    emit = getattr(ctx, 'emit', None) if ctx else None
    
    try:
        import sys
        import os
        import time
        sys.path.append(os.path.join(os.path.dirname(__file__), '..'))
        from src.services.pet_store import find_deleted_pets_ready_to_purge, remove
    except ImportError:
        if logger:
            logger.error('❌ Deletion Reaper failed - import error')
        return

    if logger:
        logger.info('🔄 Deletion Reaper started - scanning for pets to purge')

    try:
        pets_to_reap = find_deleted_pets_ready_to_purge()
        
        if not pets_to_reap:
            if logger:
                logger.info('✅ Deletion Reaper completed - no pets to purge')
            
            # No pets ready for purging
            return

        purged_count = 0
        
        for pet in pets_to_reap:
            success = remove(pet['id'])
            
            if success:
                purged_count += 1
                
                if logger:
                    logger.info('💀 Pet permanently purged', {
                        'petId': pet['id'],
                        'name': pet['name'],
                        'deletedAt': time.strftime('%Y-%m-%dT%H:%M:%S', time.gmtime(pet['deletedAt'] / 1000)),
                        'purgeAt': time.strftime('%Y-%m-%dT%H:%M:%S', time.gmtime(pet['purgeAt'] / 1000))
                    })

                # Pet purged successfully
            else:
                if logger:
                    logger.warn('⚠️ Failed to purge pet', {'petId': pet['id'], 'name': pet['name']})

        if logger:
            logger.info('✅ Deletion Reaper completed', {
                'totalScanned': len(pets_to_reap),
                'purgedCount': purged_count,
                'failedCount': len(pets_to_reap) - purged_count
            })

        # Deletion reaper completed successfully

    except Exception as error:
        if logger:
            logger.error('❌ Deletion Reaper error', {'error': str(error)})
